tests/core/test_structured_data_signing.py

```diff
@@ -34,124 +34,263 @@
+        ),
+    )
+)
+def test_encode_data(
+    example, expected_hex,
+    eip712_example_types, eip712_example_message,
+    eip712_example_with_array_types, eip712_example_with_array_message,
+):
     primary_type = "Mail"
-    expected_hex_value = (
-        "b19c5c7781083ed7325660d847f8c547b744917d67b38858b00bce7e62a4866bfc71e5fa27ff56c350aa531bc1"
-        "29ebdf613b772b6604664f5d8dbe21b85eb0c8cd54f074a4af31b4411ff6a60c9719dbd559c221c8ac3492d9d8"
-        "72b041d703d1a88b5e35e8f6bccc3d72ef467ed581615888f8cf34df19d068d04225439be65ab5aadf3154a261"
-        "abdd9086fc627b61efca26ae5702701d05cd2305f7c52a2fc8"
+    _types = (
+        eip712_example_types if example == 'eip712'
+        else eip712_example_with_array_types if example == 'eip712_with_array'
+        else pytest.fail('unreachable code')
     )
-    assert encode_data(primary_type, types, message).hex() == expected_hex_value
+    _message = (
+        eip712_example_message if example == 'eip712'
+        else eip712_example_with_array_message if example == 'eip712_with_array'
+        else pytest.fail('unreachable code')
+    )
+    assert encode_data(primary_type, _types, _message).hex() == expected_hex
 
 
-def test_hash_struct_main_message(structured_valid_data_json_string):
-    structured_data = json.loads(structured_valid_data_json_string)
-    expected_hex_value = "76649452daa3e4101beef5b01669fd0de45ece35188d529703c73526a2454521"
-    assert hash_message(structured_data).hex() == expected_hex_value
+@pytest.mark.parametrize(
+    'example, expected_hex',
+    (
+        ('eip712', 'c52c0ee5d84264471806290a3f2c4cecfc5490626bf912d01f240d7a274b371e'),
+        ('eip712_with_array', '76649452daa3e4101beef5b01669fd0de45ece35188d529703c73526a2454521'),
+    )
+)
+def test_hash_struct_main_message(
+    example, expected_hex, eip712_example_json_string, eip712_example_with_array_json_string
+):
+    structured_data = (
+        json.loads(eip712_example_json_string) if example == 'eip712'
+        else json.loads(eip712_example_with_array_json_string) if example == 'eip712_with_array'
+        else pytest.mark.fail('unreachable code')
+    )
+    assert hash_message(structured_data).hex() == expected_hex
 
 
-def test_hash_struct_domain(structured_valid_data_json_string):
-    structured_data = json.loads(structured_valid_data_json_string)
-    expected_hex_value = "f2cee375fa42b42143804025fc449deafd50cc031ca257e0b194a650a912090f"
-    assert hash_domain(structured_data).hex() == expected_hex_value
+@pytest.mark.parametrize('example', ('eip712', 'eip712_with_array',))
+def test_hash_struct_domain(
+    example, eip712_example_json_string, eip712_example_with_array_json_string
+):
+    structured_data = (
+        json.loads(eip712_example_json_string) if example == 'eip712'
+        else json.loads(eip712_example_with_array_json_string) if example == 'eip712_with_array'
+        else pytest.mark.fail('unreachable code')
```

Comment:
sometimes you use `pytest.mark.fail` and sometimes `pytest.fail`. Is there a difference?